### DISCUSSION TITLE
ref.listen + execute setState within the body of listen vs just ref.watch

### DISCUSSION THREAD

**Discussion by IoannisChadjiminas** [rrousselGit/riverpod#917] [General]
Hello,

I would like to ask whether is good practise to execute setState within the body of ref.listen instead of executing ref.watch. I have a case where I don't want to execute ref.watch but just ref.listen and then execute setState within it.

Regards,
John

**Comment by daniellampl** [comment]
It depends on why you would like to call `setState` in the `ref.listen` callback instead of using `ref.watch`. If you only want your UI to update if the provider's state matches a specific condition, I would use `.select` to explicitly listen to state changes that are matching a given condition, like so:

```
final bool roomCapacityReached = ref.watch(peopleCountProvider.select((int peopleCount) => peopleCount >= 250));
```

This will only update the widget if `roomCapacityReached` changes, instead of whenever the state of the `peopleCountProvider` changes.
